linalg/constrained.hpp

```diff
@@ -0,0 +1,272 @@
+                       Array<int>& secondary_dofs);
+
+   ~EliminationCGSolver();
+
+   void SetOperator(const Operator& op) { }
+
+   void Mult(const Vector& x, Vector& y) const;
+
+private:
+   /**
+      This assumes the primary/secondary dofs are cleanly separated in
+      the matrix, and the given index tells you where.
+
+      We want to move away from this assumption, the first step
+      is to separate its logic in this method.
+   */
+   void BuildSeparatedInterfaceDofs(int firstblocksize);
+
+   void BuildPreconditioner();
+
+   SparseMatrix& A_;
+   SparseMatrix& B_;
+   Array<int> first_interface_dofs_;
+   Array<int> second_interface_dofs_;
+   EliminationProjection * projector_;
+   HypreParMatrix * h_explicit_operator_;
+   HypreBoomerAMG * prec_;
+};
+
+/** @brief Solve constrained system with penalty method; see ConstrainedSolver.
+
+    Uses a HypreBoomerAMG preconditioner for the penalized system. Only
+    approximates the solution, better approximation with higher penalty,
+    but with higher penalty the preconditioner is less effective. */
+class PenaltyConstrainedSolver : public IterativeSolver
+{
+public:
+   PenaltyConstrainedSolver(HypreParMatrix& A, SparseMatrix& B, double penalty_);
+
+   PenaltyConstrainedSolver(HypreParMatrix& A, HypreParMatrix& B, double penalty_);
+
+   ~PenaltyConstrainedSolver();
+
+   void Mult(const Vector& x, Vector& y) const;
+
+   void SetOperator(const Operator& op) { }
+
+private:
+   void Initialize(HypreParMatrix& A, HypreParMatrix& B);
+
+   double penalty;
+   Operator& constraintB;
+   HypreParMatrix * penalized_mat;
+   HypreBoomerAMG * prec;
+};
+
+#endif
+
+/** @brief Solve constrained system by solving original mixed sysetm;
+    see ConstrainedSolver.
+
+    Solves the saddle-point problem with a block-diagonal preconditioner, with
+    user-provided preconditioner in the top-left block and an identity matrix
+    in the bottom-right. */
+class SchurConstrainedSolver : public IterativeSolver
+{
+public:
+   SchurConstrainedSolver(BlockOperator& block_op_,
+                          Solver& primal_pc_);
+   virtual ~SchurConstrainedSolver();
+
+   void SetOperator(const Operator& op) { }
+
+   void Mult(const Vector& x, Vector& y) const;
+
+private:
+   BlockOperator& block_op;
+   Solver& primal_pc;
+   BlockDiagonalPreconditioner block_pc;
+   Solver * dual_pc;  // owned
+};
+
+
+/** @brief A class to solve the constrained system \f$ Ax = f \f$
+    subject to the constraint \f$ B x = r \f$ in an abstract setting.
+
+    Although this object may not use the below formulation, for
+    understanding some of its methods and notation you can think of
+    it as solving the saddle-point system
+
+     (  A   B^T  )  ( x )         (  f  )
+     (  B        )  ( lambda)  =  (  r  )
+
+     Not to be confused with ConstrainedOperator, which is totally
+     different. */
+class ConstrainedSolver : public Solver
+{
+public:
+   ConstrainedSolver(Operator& A, Operator& B);
```

Comment:
Sam has some ideas on local constraint transformations instead of a global operator B. I think this API could still work for that, as long as B is just a generic operator. We should discuss this at some point.

Reply:
Yes we should!